### PR TITLE
Datatype leak fix, again

### PR DESCRIPTION
**Pull request by roystgnr**
This will hopefully fix a valgrind error discovered in #1857, as well
as preemptively fix the same problem with all our other remaining MPI
types.  I have no idea why I didn't do all these when I was putting
together #1616, and I'm completely baffled that the valgrind output
there didn't catch the HilbertIndices case.

@lindsayad, could you give this a try?

### CODE REVIEW

**Comment by jwpeterson** [comment]
And here I've been telling everyone for the last two days that you were already `free()`ing all the MPI committed data types!

**Comment by lindsayad** [comment]
So the previous errors are gone, but now we have new traces:
```
==3366== 40 bytes in 1 blocks are definitely lost in loss record 6 of 27                                                                                                                                                                    
==3366==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                  
==3366==    by 0x7351237: MPIR_Type_create_struct_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                           
==3366==    by 0x73519B6: PMPI_Type_create_struct (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                
==3366==    by 0x4E66FC5: PMPI_Type_create_struct (libmpiwrap.c:2727)                                                                                                                                                                       
==3366==    by 0x58E3BC3: StandardType (parallel_algebra.h:194)                                                                                                                                                                             
==3366==    by 0x58E3BC3: void libMesh::Parallel::Communicator::min<libMesh::Point>(libMesh::Point&) const (parallel_implementation.h:347)                                                                                                  
==3366==    by 0x58D5F1D: libMesh::MeshTools::create_bounding_box(libMesh::MeshBase const&) (mesh_tools.C:387)                                                                                                                              
==3366==    by 0x59B192C: libMesh::MetisPartitioner::partition_range(libMesh::MeshBase&, libMesh::MeshBase::element_iterator, libMesh::MeshBase::element_iterator, unsigned int) (metis_partitioner.C:105)                                  
==3366==    by 0x59B15BB: libMesh::MetisPartitioner::_do_partition(libMesh::MeshBase&, unsigned int) (metis_partitioner.C:500)                                                                                                              
==3366==    by 0x59C637D: libMesh::Partitioner::partition(libMesh::MeshBase&, unsigned int) (partitioner.C:86)                                                                                                                              
==3366==    by 0x57E180B: partition (mesh_base.h:750)                                                                                                                                                                                       
==3366==    by 0x57E180B: libMesh::MeshBase::prepare_for_use(bool, bool) (mesh_base.C:242)                                                                                                                                                  
==3366==    by 0x58265FA: libMesh::MeshTools::Generation::build_cube(libMesh::UnstructuredMesh&, unsigned int, unsigned int, unsigned int, double, double, double, double, double, double, libMesh::ElemType, bool) (mesh_generation.C:1455\
)                                                                                                                                                                                                                                           
==3366==    by 0x413A69: main (vector_fe_ex3.C:88)               

==3366== 72 bytes in 1 blocks are definitely lost in loss record 13 of 27                                                                                                                                                                   
==3366==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                  
==3366==    by 0x73F216F: MPIR_Dataloop_alloc_and_copy (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                           
==3366==    by 0x73F4BAB: MPIR_Dataloop_create_contiguous (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                        
==3366==    by 0x73F43D7: MPIR_Dataloop_create_blockindexed (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                      
==3366==    by 0x73F6952: MPIR_Dataloop_create_struct (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                            
==3366==    by 0x73F36BD: MPIR_Dataloop_create (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                   
==3366==    by 0x7345FFF: MPIR_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                       
==3366==    by 0x73460EE: MPIR_Type_commit_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                  
==3366==    by 0x7346338: PMPI_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                       
==3366==    by 0x4E468F0: PMPI_Type_commit (libmpiwrap.c:1769)                                                                                                                                                                              
==3366==    by 0x58E3BCB: StandardType (parallel_algebra.h:197)                                                                                                                                                                             
==3366==    by 0x58E3BCB: void libMesh::Parallel::Communicator::min<libMesh::Point>(libMesh::Point&) const (parallel_implementation.h:347)                                                                                                  
==3366==    by 0x58D5F1D: libMesh::MeshTools::create_bounding_box(libMesh::MeshBase const&) (mesh_tools.C:387)      
```

**Comment by lindsayad** [comment]
Something that's interesting is that we see this same kind of trace coming out of PETSc:
```
==3366== 56 bytes in 1 blocks are definitely lost in loss record 8 of 27                                                                                                                                                                    
==3366==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                  
==3366==    by 0x7351237: MPIR_Type_create_struct_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                           
==3366==    by 0x73519B6: PMPI_Type_create_struct (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                
==3366==    by 0x4E66FC5: PMPI_Type_create_struct (libmpiwrap.c:2727)                                                                                                                                                                       
==3366==    by 0x888CE35: MatStashBlockTypeSetUp (matstash.c:785)                                                                                                                                                                           
==3366==    by 0x888DD83: MatStashScatterBegin_BTS (matstash.c:853)                                                                                                                                                                         
==3366==    by 0x8889892: MatStashScatterBegin_Private (matstash.c:462)                                                                                                                                                                     
==3366==    by 0x87C844C: MatAssemblyBegin_MPIAIJ (mpiaij.c:774)                                                                                                                                                                            
==3366==    by 0x82D0524: MatAssemblyBegin (matrix.c:5247)                                                                                                                                                                                  
==3366==    by 0x597F6A4: libMesh::PetscMatrix<double>::close() (petsc_matrix.C:972)                                                                                                                                                        
==3366==    by 0x5AEA954: libMesh::PetscLinearSolver<double>::solve(libMesh::SparseMatrix<double>&, libMesh::SparseMatrix<double>&, libMesh::NumericVector<double>&, libMesh::NumericVector<double>&, double, unsigned int) (petsc_linear_s\
olver.C:455)                                                                                                                                                                                                                                
==3366==    by 0x5AD0E55: libMesh::PetscLinearSolver<double>::solve(libMesh::SparseMatrix<double>&, libMesh::NumericVector<double>&, libMesh::NumericVector<double>&, double, unsigned int) (petsc_linear_solver.h:163)  

==3366== 72 bytes in 1 blocks are definitely lost in loss record 17 of 27                                                                                                                                                                   
==3366==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)                                                                                                                                                                                  
==3366==    by 0x73F216F: MPIR_Dataloop_alloc_and_copy (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                           
==3366==    by 0x73F4BAB: MPIR_Dataloop_create_contiguous (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                        
==3366==    by 0x73F55B2: MPIR_Dataloop_create_indexed (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                           
==3366==    by 0x73F6C5A: MPIR_Dataloop_create_struct (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                            
==3366==    by 0x73F36BD: MPIR_Dataloop_create (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                   
==3366==    by 0x7345FFF: MPIR_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                       
==3366==    by 0x73460EE: MPIR_Type_commit_impl (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                  
==3366==    by 0x7346338: PMPI_Type_commit (in /ssd/lindad/mpich-3.3b1/installed/lib/libmpi.so.0.0.0)                                                                                                                                       
==3366==    by 0x4E468F0: PMPI_Type_commit (libmpiwrap.c:1769)                                                                                                                                                                              
==3366==    by 0x888CE96: MatStashBlockTypeSetUp (matstash.c:786)                                                                                                                                                                           
==3366==    by 0x888DD83: MatStashScatterBegin_BTS (matstash.c:853)                          
```

So a `PMPI_Type_create_struct` followed by a `PMPI_Type_commit`. What @fdkong and I saw is that if we run a PETSc only example which hits the same code path, we don't see any leaks.

**Comment by roystgnr** [comment]
The new traces are my fault - I forgot we now need a copy constructor too not just a destructor.

**Comment by roystgnr** [comment]
Hmm... no, that would risk the opposite problem, actually.  Now I'm not sure what's going on with the parallel_algebra.h StandardTypes...

**Comment by roystgnr** [comment]
@lindsayad, can you get any of these leaks to trigger with any of the libMesh examples?  Possibly in a --enable-parmesh build at least?

**Comment by roystgnr** [comment]
I'm an idiot.  You're using vector_fe_ex3...

**Comment by lindsayad** [comment]
Every single error follows an `MPI_Type_create_struct`, both in PETSc and LibMesh, specifically **every** call to `MPI_Type_create_struct` yields an error. This is mysterious considering that I can run PETSc examples which call paths (including one path that is the same as that called through LibMesh!) that contain `MPI_Type_create_struct` and get valgrind clean results. Is this a C vs C++ issue...? I dunno, I'm really starting to think these remaining errors might be red herrings.

**Comment by lindsayad** [comment]
@jedbrown You authored the `matstash.c` code...do you have any idea how we might generate a leak when calling through LibMesh, but not see anything when working solely in PETSc?

**Comment by jedbrown** [comment]
Perhaps there is an edge case due to a particular sequence of assemblies. What is a minimal test case I can use to reproduce?

**Comment by roystgnr** [comment]
> Is this a C vs C++ issue...?

No, but that's not the only difference in how we vs they use MPI.

... wild guess, but can you make sure that libMesh is calling MPI_Finalize properly?

**Comment by lindsayad** [comment]
Pretend that last post never happened...that was with one element. With two elements (and two procs), the leaks come back. @jedbrown here is a [gist](https://gist.github.com/lindsayad/8ca858a52cacff623a4429a0963067dd) with a "minimum" working example

@roystgnr I'll take a look at Finalize

**Comment by lindsayad** [comment]
I don't see any problems with the call to `MPI_Finalize`...I stepped through, saw that `MPI_Finalize` does *not* get called from PETSc; it *does* get called in the `LibMeshInit` destructor. But honestly, I'm not an MPI expert.

**Comment by fdkong** [comment]
@jedbrown, Thanks so much for your help on this. It is definitely better to let `DMRegister` handle `XXXInitializePackage` if necessary.

At the libMesh side, it might be better to not call  `DMRegister` if we are not going to use `DM_libMesh`. It might be better to move `DMRegister` to the right place where we are actually going to create a `DM_libMesh`.  Most users do not use `DM_libMesh`

**Comment by lindsayad** [comment]
@jedbrown How are you leak testing? I applied the one-line patch and I'm still seeing the leaks out of `MatStash`.

**Comment by fdkong** [comment]
@lindsayad Could you also try to remove `DMRegister`? And then `valgrind` again with this change. I just want to confirm our conclusion.

**Comment by lindsayad** [comment]
I also ran without `DMRegister` and still saw the same leaks

**Comment by jedbrown** [comment]
BTW, the `DMRegister` stuff will be fixed once https://bitbucket.org/petsc/petsc/pull-requests/1125/barry-fix-xxxregister-memory-leaks-maint/diff is merged.

**Comment by lindsayad** [comment]
Yes this is with MPICH. So you don't even see the libMesh only leaks that I see? I would be happy to be wrong..
```
$ mpiexec --version
HYDRA build details:
    Version:                                 3.3b1
    Release Date:                            Mon Feb  5 10:16:15 CST 2018
    CC:                              gcc    
    CXX:                             g++    
    F77:                             gfortran   
    F90:                             gfortran   
    Configure options:                       '--disable-option-checking' '--prefix=/ssd/lindad/mpich-3.3b1/installed' '--cache-file=/dev/null' '--srcdir=../../../../src/pm/hydra' 'CC=gcc' 'CFLAGS= -O2' 'LDFLAGS=' 'LIBS=' 'CPPFLAGS= -I/ssd/lindad/mpich-3.3b1/build/src/mpl/include -I/ssd/lindad/mpich-3.3b1/src/mpl/include -I/ssd/lindad/mpich-3.3b1/src/openpa/src -I/ssd/lindad/mpich-3.3b1/build/src/openpa/src -D_REENTRANT -I/ssd/lindad/mpich-3.3b1/build/src/mpi/romio/include' 'MPLLIBNAME=mpl'
    Process Manager:                         pmi
    Launchers available:                     ssh rsh fork slurm ll lsf sge manual persist
    Topology libraries available:            hwloc
    Resource management kernels available:   user slurm ll lsf sge pbs cobalt
    Checkpointing libraries available:       
    Demux engines available:                 poll select

$ valgrind --version
valgrind-3.13.0
```
MPICH was downloaded from the same URL as that specified in the PETSc configure. Valgrind is from the 3.13.0 tarball off their download page.

I've also reproduced the same set of leaks with a combination of MPICH 3.2.1 and valgrind 3.14.0.GIT (e.g. yesterday's build of valgrind master)

**Comment by jedbrown** [comment]
Hmm, interesting. Just to control variables, what compiler are you using? Can you reproduce with a PETSc example?
```
make -f gmakefile $PETSC_ARCH/tests/snes/examples/tutorials/ex12
mpiexec -n 2 valgrind --track-origins=yes --leak-check=full $PETSC_ARCH/tests/snes/examples/tutorials/ex12 -refinement_limit 0.0 -simplex 0 -interpolate -bc_type dirichlet -petscspace_degree 1 -cells 2,2
```

**Comment by lindsayad** [comment]
Configuration 1 on our group's Linux workhorse (Ubuntu 16.04.5):
- gcc 5.4.0
- MPICH 3.3b1
- valgrind 3.13.0 (from tar ball)

Configuration 2 on virtual machine on personal laptop (Ubuntu 18.04.1):
- gcc 7.3.0
- MPICH 3.2.1
- valgrind 3.14.0 (built from commit 97365bada)

**Comment by lindsayad** [comment]
I cannot reproduce with the PETSc example. Everything is clean

**Comment by jedbrown** [comment]
Hmm, my tests have been with gcc-8.2.1, mpich-3.3b1 and ompi-3.1.1, and recent valgrind as mentioned. I can spin up a VM to attempt to match your configuration, but it's a bit time-consuming. The (weird) warnings above that related to matstash are present for me with both the libmesh example and that PETSc example. This is with petsc-3.10 and nearby commits.

**Comment by lindsayad** [comment]
I'll try with your config and get back to you

**Comment by lindsayad** [comment]
So @jedbrown I guess that you do not run with [mpiwrap](http://valgrind.org/docs/manual/mc-manual.html#mc-manual.mpiwrap)? I guess the purpose of that library was to reduce memory leak false positives out of mpi...but it seems that here it has done the opposite. If I run without `LD_PRELOAD=/path/to/libmpiwrap`, then I do not see `matstash` leaks.

@roystgnr with this additional patch:
```
diff --git a/include/parallel/parallel_algebra.h b/include/parallel/parallel_algebra.h
index f3e2b70..7171120 100644
--- a/include/parallel/parallel_algebra.h
+++ b/include/parallel/parallel_algebra.h
@@ -216,6 +216,9 @@ public:
 
     libmesh_call_mpi
       (MPI_Type_commit (&_datatype));
+
+    libmesh_call_mpi
+      (MPI_Type_free(&tmptype));
 #endif // #ifdef LIBMESH_HAVE_MPI
   }
 
```
then I think we are totally clean, at least on the MWE.

**Comment by jwpeterson** [comment]
FYI @roystgnr, I pushed @lindsayad's fix directly to your branch.

**Comment by jedbrown** [comment]
@lindsayad Yes, I was skipping mpiwrap because it is built for Open MPI on my machine, but I usually use MPICH for testing (at least when valgrind-cleanliness is important). Unfortunately, this is a pretty complex scenario to file a bug report to Valgrind about.

File: include/parallel/parallel_algebra.h [hunk 1]
```diff
@@ -216,6 +216,9 @@ class StandardType<Point> : public DataType
 
     libmesh_call_mpi
       (MPI_Type_commit (&_datatype));
+
+    libmesh_call_mpi
+      (MPI_Type_free(&tmptype));
```
Comment (roystgnr) [inline include/parallel/parallel_algebra.h]: We (ok, I) have this same bug in two other places in this file, right?

**Comment by lindsayad** [comment]
@jwpeterson could we apply this patch as well (as pointed out by @roystgnr)?
```
diff --git a/include/parallel/parallel_algebra.h b/include/parallel/parallel_algebra.h
index 717112030..b0d7bd1a7 100644
--- a/include/parallel/parallel_algebra.h
+++ b/include/parallel/parallel_algebra.h
@@ -91,6 +91,9 @@ public:
 
     libmesh_call_mpi
       (MPI_Type_commit (&_datatype));
+
+    libmesh_call_mpi
+      (MPI_Type_free (&tmptype));
 #endif // #ifdef LIBMESH_HAVE_MPI
   }
 
@@ -151,6 +154,9 @@ public:
 
     libmesh_call_mpi
       (MPI_Type_commit (&_datatype));
+
+    libmesh_call_mpi
+      (MPI_Type_free (&tmptype));
 #endif // #ifdef LIBMESH_HAVE_MPI
   }
 
@@ -218,7 +224,7 @@ public:
       (MPI_Type_commit (&_datatype));
 
     libmesh_call_mpi
-      (MPI_Type_free(&tmptype));
+      (MPI_Type_free (&tmptype));
 #endif // #ifdef LIBMESH_HAVE_MPI
   }
 
```
